README.md
# NAME

Plack::Middleware::Signposting - a base class for Plack implementations of the [Signposting](https://signposting.org) protocol

![Test status](https://github.com/LibreCat/Plack-Middleware-Signposting/actions/workflows/linux.yml/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/LibreCat/Plack-Middleware-Signposting/badge.svg?branch=master)](https://coveralls.io/github/LibreCat/Plack-Middleware-Signposting?branch=master)
[![CPANTS kwalitee](http://cpants.cpanauthors.org/dist/Plack-Middleware-Signposting.png)](http://cpants.cpanauthors.org/dist/Plack-Middleware-Signposting)

# SYNOPSIS

    package Plack::Middleware::Signposting::Foo;

    use Moo;

    extends 'Plack::Middleware::Signposting';

    sub call {
        my ($self, $env) = @_;

        ...
        my @data = ("0001", $relation, $type);
        $self->to_link_format(\@data);
    }

# METHODS

- to\_link\_format(\\@ARRAY)

    This method produces the format for the link header.

# MODULES

- [Plack::Middleware::Signposting::JSON](https://metacpan.org/pod/Plack%3A%3AMiddleware%3A%3ASignposting%3A%3AJSON)
- [Plack::Middleware::Signposting::Catmandu](https://metacpan.org/pod/Plack%3A%3AMiddleware%3A%3ASignposting%3A%3ACatmandu)

# AUTHOR

Vitali Peil, `<vitali.peil at uni-bielefeld.de>`

Nicolas Steenlant, `<nicolas.steenlant at ugent.be>`

# CONTRIBUTORS

Mohammad S Anwar (@manwar)

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

[Plack::Middleware](https://metacpan.org/pod/Plack%3A%3AMiddleware)


Changes
Revision history for Plack-Middleware-Signposting

{{$NEXT}}

0.06  2022-06-27 11:30:57 CEST
    - use github actions
    - refresh distro

0.05  2018-02-14 13:44:56 CET
    - more POD
    - better test

0.04  2018-02-09 12:53:04 CET
    - correct path matching
    - better pod

0.03  2018-01-26 12:52:49 CET
	- introdcue sub class 'Catmandu'
	- add pod
	- add tests

0.02  2017-05-29 21:18:07 CEST
    - add base class

0.01  2017-05-29 20:43:01 CEST
    - Initial release


lib/Plack/Middleware/Signposting/JSON.pm
package Plack::Middleware::Signposting::JSON;

our $VERSION = '0.04';

use Catmandu::Sane;
use Catmandu;
use Catmandu::Fix;
use JSON qw(decode_json);
use Plack::Request;
use Plack::Util::Accessor qw(fix);
use Moo;

extends 'Plack::Middleware::Signposting';

sub call {
    my ($self, $env) = @_;

    my $request = Plack::Request->new($env);
    my $res = $self->app->($env);

    # only get/head requests
    return $res unless $request->method =~ m{^get|head$}i;

    # see http://search.cpan.org/~miyagawa/Plack-1.0044/lib/Plack/Middleware.pm#RESPONSE_CALLBACK
    return $self->response_cb($res, sub {
        my $res = shift;

        my $content_type = Plack::Util::header_get($res->[1], 'Content-Type') || '';
        # only json responses
        return unless $content_type =~ m{^application/json|application\/vnd\.api\+json}i;
        # ignore streaming response for now
        return unless ref $res->[2] eq 'ARRAY';

        my $body = join('', @{$res->[2]});
        my $data = decode_json($body);

        if (ref $data && ref $data eq 'ARRAY') {
            $data = $data->[0];
        }

        my $fix = $self->fix ? $self->fix : 'nothing()';
        my $fixer = Catmandu::Fix->new(fixes => [$fix]);
        $fixer->fix($data);

        # add information to the 'Link' header
        if ($data->{signs}) {
            Plack::Util::header_push(
                $res->[1],
                'Link' => $self->to_link_format( @{$data->{signs}} )
            );
        }
    });
}

1;

__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::Signposting::JSON - A Signposting implementation for JSON content

=head1 SYNOPSIS

    my $json_string = '{"hello":"world",....}';
    builder {
       enable "Plack::Middleware::Signposting::JSON";

       sub { 200, [ 'Content-Type' => 'text/plain' ], [ $json_string ] };
    };

=head1 SEE ALSO

L<Plack::Middleware>

=cut


lib/Plack/Middleware/Signposting/Catmandu.pm
package Plack::Middleware::Signposting::Catmandu;

our $VERSION = '0.04';

use Catmandu::Sane;
use Catmandu;
use Catmandu::Fix;
use Plack::Request;
use Plack::Util::Accessor;
use Moo;

extends 'Plack::Middleware::Signposting';

has store => (is => 'ro');
has bag => (is => 'ro');
has _bag => (is => 'lazy');
has fix => (is => 'ro');
has match_paths => (is => 'ro');
has _fixer => (is => 'lazy');

sub _build__bag {
    my ($self) = @_;
    Catmandu->store($self->store)->bag($self->bag);
}

sub _build__fixer {
    my ($self) = @_;
    Catmandu::Fix->new(fixes => [$self->fix]);
}

sub call {
    my ($self, $env) = @_;

    my $request = Plack::Request->new($env);
    my $res = $self->app->($env);

    my $bag = $self->_bag;
    my $fixer = $self->_fixer;

    # only get/head requests
    return $res unless $request->method =~ m{^get|head$}i;

    my $id;
    my $match_paths = $self->match_paths;
    foreach my $p (@$match_paths) {
        if ($request->path =~ /$p/) {
            $id = $1;
            last;
        }
    }

    return $res unless $id;

    # see http://search.cpan.org/~miyagawa/Plack-1.0044/lib/Plack/Middleware.pm#RESPONSE_CALLBACK
    return $self->response_cb($res, sub {
        my $res = shift;

        # ignore streaming response for now
        return unless ref $res->[2] eq 'ARRAY';

        my $data = $bag->get($id) || return;

        $fixer->fix($data);

        # add information to the 'Link' header
        if ($data->{signs}) {
            Plack::Util::header_push(
                $res->[1],
                'Link' => $self->to_link_format( @{$data->{signs}} )
            );
        }
    });
}

1;

__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::Signposting::Catmandu - A Signposting implementation from a Catmandu store

=head1 SYNOPSIS

    builder {
        enable "Plack::Middleware::Signposting::Catmandu",
            store => 'library',
            bag => 'books',
            fix => 'signs.fix', #optional
            match_paths => ['publication/(\w+)/*', 'record/(\w+)/*'],
            ;

        # ...
    };

=head1 SEE ALSO

L<Plack::Middleware>, L<Catmandu>

=cut


lib/Plack/Middleware/Signposting.pm
package Plack::Middleware::Signposting;

our $VERSION = '0.05';

use Catmandu::Sane;
use parent 'Plack::Middleware';

sub to_link_format {
    my ($self, @signs) = @_;

    my $body = join(", ", map {
        my ($uri, $relation, $type) = @$_;
        my $link_text = qq|<$uri>; rel="$relation"|;
        $link_text .= qq|; type="$type"| if $type;
        $link_text;
    } @signs);

    $body;
}

1;

__END__

=encoding utf-8

=head1 NAME

Plack::Middleware::Signposting - a base class for Plack implementations of the L<Signposting|https://signposting.org> protocol

=begin markdown

![Test status](https://github.com/LibreCat/Plack-Middleware-Signposting/actions/workflows/linux.yml/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/LibreCat/Plack-Middleware-Signposting/badge.svg?branch=master)](https://coveralls.io/github/LibreCat/Plack-Middleware-Signposting?branch=master)
[![CPANTS kwalitee](http://cpants.cpanauthors.org/dist/Plack-Middleware-Signposting.png)](http://cpants.cpanauthors.org/dist/Plack-Middleware-Signposting)

=end markdown

=head1 SYNOPSIS

    package Plack::Middleware::Signposting::Foo;

    use Moo;

    extends 'Plack::Middleware::Signposting';

    sub call {
        my ($self, $env) = @_;

        ...
        my @data = ("0001", $relation, $type);
        $self->to_link_format(\@data);
    }

=head1 METHODS

=over

=item * to_link_format(\@ARRAY)

This method produces the format for the link header.

=back

=head1 MODULES

=over

=item * L<Plack::Middleware::Signposting::JSON>

=item * L<Plack::Middleware::Signposting::Catmandu>

=back

=head1 AUTHOR

Vitali Peil, C<< <vitali.peil at uni-bielefeld.de> >>

Nicolas Steenlant, C<< <nicolas.steenlant at ugent.be> >>

=head1 CONTRIBUTORS

Mohammad S Anwar (@manwar)

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=head1 SEE ALSO

L<Plack::Middleware>

=cut
